File: pkg/soap/enrollment.go
package soap

import (
	"encoding/base64"
	"fmt"

	"github.com/mattrax/xml"
)

// EnrollmentRequest contains the device information and identity certificate CSR
type EnrollmentRequest struct {
	XMLName xml.Name      `xml:"s:Envelope"`
	Header  RequestHeader `xml:"s:Header"`
	Body    struct {
		TokenType           string              `xml:"wst:TokenType"`
		RequestType         string              `xml:"wst:RequestType"`
		BinarySecurityToken BinarySecurityToken `xml:"wsse:BinarySecurityToken"`
		AdditionalContext   []ContextItem       `xml:"ac:AdditionalContext>ac:ContextItem"`
	} `xml:"s:Body>wst:RequestSecurityToken"`
}

// BinarySecurityToken contains the CSR for the request and wap-provisioning payload for the response
type BinarySecurityToken struct {
	ValueType    string `xml:"ValueType,attr"`
	EncodingType string `xml:"EncodingType,attr"`
	Value        string `xml:",chardata"`
}

// DecodedValue decodes the binary security token based off the specified encoding type
func (bst BinarySecurityToken) DecodedValue() ([]byte, error) {
	if bst.EncodingType == "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd#base64binary" {
		return base64.StdEncoding.DecodeString(bst.Value)
	}
	return nil, fmt.Errorf("encoding not supported")
}

// EnrollmentRequestTypeIssue is the RequestType for the Enrollment request body used to issue a new certificate
const EnrollmentRequestTypeIssue = "http://docs.oasis-open.org/ws-sx/ws-trust/200512/Issue"

// EnrollmentRequestTypeRenew is the RequestType for the Certificate Renewal request body to renew an existing certificate
const EnrollmentRequestTypeRenew = "http://docs.oasis-open.org/ws-sx/ws-trust/200512/Renew"

// BinarySecurityTokenTypePKCS7 is the BinarySecurityToken ValueType for a PKCS7 which is used for the Renew request type
const BinarySecurityTokenTypePKCS7 = "http://schemas.microsoft.com/windows/pki/2009/01/enrollment#PKCS7"

// BinarySecurityTokenTypePKCS10 is the BinarySecurityToken ValueType for a PKCS10 which is used for the Issue request type
const BinarySecurityTokenTypePKCS10 = "http://schemas.microsoft.com/windows/pki/2009/01/enrollment#PKCS10"

// ContextItem are key value pairs which contains information about the device being enrolled
type ContextItem struct {
	Name  string `xml:"Name,attr"`
	Value string `xml:"ac:Value"`
}

// GetAdditionalContextItem retrieves the first AdditionalContext item with the specified name
func (cmd EnrollmentRequest) GetAdditionalContextItem(name string) string {
	for _, contextItem := range cmd.Body.AdditionalContext {
		if contextItem.Name == name {
			return contextItem.Value
		}
	}
	return ""
}

// GetAdditionalContextItems retrieves all AdditionalContext items with a specified name
func (cmd EnrollmentRequest) GetAdditionalContextItems(name string) []string {
	var contextItems []string
	for _, contextItem := range cmd.Body.AdditionalContext {
		if contextItem.Name == name {
			contextItems = append(contextItems, contextItem.Value)
		}
	}
	return contextItems
}

// EnrollmentResponse contains the management client configuration and signed identity certificate
type EnrollmentResponse struct {
	XMLName xml.Name `xml:"http://docs.oasis-open.org/ws-sx/ws-trust/200512 RequestSecurityTokenResponseCollection"`
	// BinarySecurityToken          BinarySecurityToken `xml:"RequestSecurityTokenResponse>TokenType"`
	TokenType                    string              `xml:"RequestSecurityTokenResponse>TokenType"`
	DispositionMessage           DispositionMessage  `xml:"http://schemas.microsoft.com/windows/pki/2009/01/enrollment RequestSecurityTokenResponse>DispositionMessage"`
	RequestedBinarySecurityToken BinarySecurityToken `xml:"http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd RequestSecurityTokenResponse>RequestedSecurityToken>BinarySecurityToken"`
	RequestID                    int                 `xml:"http://schemas.microsoft.com/windows/pki/2009/01/enrollment RequestSecurityTokenResponse>RequestID"`
}

// DispositionMessage is an extension to the string type that allows an attribute definition of the language
type DispositionMessage struct {
	Lang  string `xml:"xml:lang,attr"`
	Value string `xml:",innerxml"`
}

// NewIssueEnrollmentResponse creates a generic enrollment response envelope for device enrollment
func NewIssueEnrollmentResponse(relatesTo string, rawProvisioningProfile []byte) ResponseEnvelope {
	var res = ResponseEnvelope{
		Header: ResponseHeader{
			RelatesTo: relatesTo,
		},
		Body: ResponseEnvelopeBody{
			Body: EnrollmentResponse{
				TokenType:          "http://schemas.microsoft.com/5.0.0.0/ConfigurationManager/Enrollment/DeviceEnrollmentToken",
				DispositionMessage: DispositionMessage{},
				RequestedBinarySecurityToken: BinarySecurityToken{
					ValueType:    "http://schemas.microsoft.com/5.0.0.0/ConfigurationManager/Enrollment/DeviceEnrollmentProvisionDoc",
					EncodingType: "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd#base64binary",
					Value:        base64.StdEncoding.EncodeToString(append([]byte(`<?xml version="1.0" encoding="UTF-8"?>`), rawProvisioningProfile...)),
				},
				RequestID: 0,
			},
		},
	}
	res.Populate("http://schemas.microsoft.com/windows/pki/2009/01/enrollment/RSTRC/wstep")
	return res
}

File: crates/ms-mde/src/enrollment.rs
use std::str::FromStr;

use easy_xml_derive::{XmlDeserialize, XmlSerialize};

use crate::{header::BinarySecurityToken, RequestHeader, ResponseHeader};

pub const ENROLLMENT_ACTION_REQUEST: &str =
    "http://schemas.microsoft.com/windows/pki/2009/01/enrollment/RST/wstep";
pub const ENROLLMENT_ACTION_RESPONSE: &str =
    "http://schemas.microsoft.com/windows/pki/2009/01/enrollment/RSTRC/wstep";

// contains the OID for the Microsoft certificate extension which includes the MDM DeviceID
pub const MICROSOFT_DEVICE_ID_EXTENSION: &[u64] = &[1, 3, 6, 1, 4, 1, 311, 66, 1, 0];

// is the RequestType for the Enrollment request body used to issue a new certificate
pub const ENROLLMENT_REQUEST_TYPE_ISSUE: &str =
    "http://docs.oasis-open.org/ws-sx/ws-trust/200512/Issue";

/// is the RequestType for the Certificate Renewal request body to renew an existing certificate
pub const ENROLLMENT_REQUEST_TYPE_RENEW: &str =
    "http://docs.oasis-open.org/ws-sx/ws-trust/200512/Renew";

/// is the BinarySecurityToken ValueType for a PKCS7 which is used for the Renew request type
pub const BINARY_SECURITY_TOKEN_TYPE_PKCS7: &str =
    "http://schemas.microsoft.com/windows/pki/2009/01/enrollment#PKCS7";

/// is the BinarySecurityToken ValueType for a PKCS10 which is used for the Issue request type
pub const BINARY_SECURITY_TOKEN_TYPE_PKCS10: &str =
    "http://schemas.microsoft.com/windows/pki/2009/01/enrollment#PKCS10";

pub const REQUEST_SECURITY_TOKEN_TYPE: &str =
    "http://schemas.microsoft.com/5.0.0.0/ConfigurationManager/Enrollment/DeviceEnrollmentToken";

pub const REQUEST_SECURITY_TOKEN_RESPONSE_COLLECTION: &str =
    "http://docs.oasis-open.org/ws-sx/ws-trust/200512";

pub const WSSE_NAMESPACE: &str =
    "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd";

///  contains the device information and identity certificate CSR
#[derive(Debug, Clone, PartialEq, Eq, Hash, XmlDeserialize, XmlSerialize)]
#[easy_xml(root, rename = "Envelope", prefix="s", namespace = {
    "a" : "http://www.w3.org/2005/08/addressing",
    "s" : "http://www.w3.org/2003/05/soap-envelope",
    "u": "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd",
    "wsse": WSSE_NAMESPACE,
    "wst": "http://docs.oasis-open.org/ws-sx/ws-trust/200512",
    "ac": "http://schemas.microsoft.com/windows/pki/2009/01/enrollment",
})]
pub struct EnrollmentRequest {
    #[easy_xml(rename = "Header", prefix = "s")]
    pub header: RequestHeader,
    #[easy_xml(rename = "Body", prefix = "s")]
    pub body: EnrollmentRequestBody,
}

impl FromStr for EnrollmentRequest {
    type Err = easy_xml::de::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        easy_xml::de::from_str(s)
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, XmlDeserialize, XmlSerialize)]
#[easy_xml(rename = "Body", prefix = "s")]
pub struct EnrollmentRequestBody {
    #[easy_xml(rename = "RequestSecurityToken", prefix = "wst")]
    pub request_security_token: EnrollmentRequestBodyRequestSecurityToken,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, XmlDeserialize, XmlSerialize)]
#[easy_xml(rename = "Body", prefix = "s")]
pub struct EnrollmentRequestBodyRequestSecurityToken {
    #[easy_xml(rename = "TokenType", prefix = "wst")]
    pub token_type: String,
    #[easy_xml(rename = "RequestType", prefix = "wst")]
    pub request_type: String,
    #[easy_xml(rename = "BinarySecurityToken", prefix = "wsse")]
    pub binary_security_token: BinarySecurityToken,
    #[easy_xml(rename = "AdditionalContext", prefix = "ac")]
    pub additional_context: AdditionalContext,
}
/// are key value pairs which contains information about the device being enrolled
#[derive(Debug, Clone, PartialEq, Eq, Hash, XmlDeserialize, XmlSerialize)]
#[easy_xml(rename = "AdditionalContext", prefix = "ac")]
pub struct AdditionalContext {
    #[easy_xml(rename = "ContextItem", prefix = "ac")]
    pub context_item: Vec<ContextItem>,
}

impl AdditionalContext {
    /// get the value of a context item by name.
    ///
    /// If multiple items with the same name exist, the first one is returned.
    pub fn get(&self, name: &str) -> Option<&str> {
        self.context_item
            .iter()
            .find(|item| item.name == name)
            .map(|item| item.value.as_str())
    }
}

/// are key value pairs which contains information about the device being enrolled
#[derive(Debug, Clone, PartialEq, Eq, Hash, XmlDeserialize, XmlSerialize)]
#[easy_xml(rename = "ContextItem", prefix = "ac")]
pub struct ContextItem {
    #[easy_xml(rename = "Name", attribute)]
    pub name: String,
    #[easy_xml(rename = "Value", prefix = "ac")]
    pub value: String,
}

// contains the management client configuration and signed identity certificate
#[derive(Debug, Clone, PartialEq, Eq, Hash, XmlDeserialize, XmlSerialize)]
#[easy_xml(root, rename = "Envelope", prefix="s",
    namespace = {
        "a" : "http://www.w3.org/2005/08/addressing",
        "s" : "http://www.w3.org/2003/05/soap-envelope",
        "u": "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd"
    }
)]
pub struct EnrollmentResponse {
    #[easy_xml(rename = "Header", prefix = "s")]
    pub header: ResponseHeader,
    #[easy_xml(rename = "Body", prefix = "s")]
    pub body: EnrollmentResponseBody,
}

impl EnrollmentResponse {
    pub fn to_string(&self) -> Result<String, easy_xml::se::Error> {
        easy_xml::se::to_string(self)
            .map(|v| v.replace(r#"<?xml version="1.0" encoding="UTF-8"?>"#, ""))
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, XmlDeserialize, XmlSerialize)]
#[easy_xml(rename = "Body", prefix = "s")]
pub struct EnrollmentResponseBody {
    #[easy_xml(rename = "RequestSecurityTokenResponseCollection")]
    pub request_security_token_response_collection: RequestSecurityTokenResponseCollection,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, XmlDeserialize, XmlSerialize)]
#[easy_xml(rename = "RequestSecurityTokenResponseCollection")]
pub struct RequestSecurityTokenResponseCollection {
    // TODO: Make this work with easy_xml properly
    #[easy_xml(rename = "xmlns", attribute)]
    pub xmlns: String,

    #[easy_xml(rename = "RequestSecurityTokenResponse")]
    pub request_security_token_response: RequestSecurityTokenResponse,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, XmlDeserialize, XmlSerialize)]
#[easy_xml(rename = "RequestSecurityTokenResponse")]
pub struct RequestSecurityTokenResponse {
    #[easy_xml(rename = "TokenType")]
    pub token_type: String,
    // <DispositionMessage xmlns="http://schemas.microsoft.com/windows/pki/2009/01/enrollment"></DispositionMessage>
    // <RequestID xmlns="http://schemas.microsoft.com/windows/pki/2009/01/enrollment">0</RequestID>
    #[easy_xml(rename = "RequestedSecurityToken")]
    pub requested_security_token: RequestedSecurityToken,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, XmlDeserialize, XmlSerialize)]
pub struct RequestedSecurityToken {
    #[easy_xml(rename = "BinarySecurityToken")]
    pub binary_security_token: BinarySecurityToken,
}
